### PR TITLE
Issue #1323 - Add support for ':of-type' modifier

### PR DESCRIPTION
**Pull request by tbieste**
Issue #1323 - Add support for ':of-type' modifier

Signed-off-by: Troy Biesterfeld <tbieste@us.ibm.com>

### CODE REVIEW

**Comment by tbieste** [comment]
Here's what the extracted parameters look like in the x_TOKEN_VALUES_V views.

| PARAMETER_NAME_ID | CODE_SYSTEM_ID | TOKEN_VALUE | COMPOSITE_ID |
| ----------------------- | ------------------- | --------------- | ------------- |
| 1191 | 20003 | 333333 | |
| 20000 | 20005 | MR1 | 0 |
| 20001 | 20010 | 333333 | 0 |
| 20000 | 20004 | MR2 | 1 |
| 20001 | 20010 | 333333 | 1 |
| 1191 | 20003 | 4444444 | | 
| 20000 | 20009 | MR3 | 2 |
| 20001 | 20010 | 4444444 | 2 |
| 20000 | 20008 | MR4 | 3 |
| 20001 | 20010 | 4444444 | 3 |

**Comment by tbieste** [comment]
This PR is dependent on changes in the PR for issue #773.

**Comment by lmsurpre** [comment]
> This PR is dependent on changes in the PR for issue #773.
> So review of this PR can focus on just the commits for this issue #1323.

Tip:  you can actually have this pull request target the branch you used for the other PR...then the diff will only show the delta.   If the other PR gets merged, github will automatically change this target to `main`.  Mind if I change the base of the PR?

**Comment by tbieste** [comment]
> > This PR is dependent on changes in the PR for issue #773.
> > So review of this PR can focus on just the commits for this issue #1323.
> 
> Tip: you can actually have this pull request target the branch you used for the other PR...then the diff will only show the delta. If the other PR gets merged, github will automatically change this target to `main`. Mind if I change the base of the PR?

Ah, ok.  I didn't realize it worked that way.  Sure, go ahead!

File: docs/src/pages/Conformance.md [hunk 1]
```diff
@@ -174,7 +174,7 @@ FHIR search modifiers are described at https://www.hl7.org/fhir/R4/search.html#m
 
 Due to performance implications, the `:exact` modifier should be used for String searches where possible.
 
-The `:text`, `:above`, `:below`, `:in`, `:not-in`, and `:of-type` modifiers are not supported in this version of the IBM FHIR server and use of these modifiers will result in an HTTP 400 error with an OperationOutcome that describes the failure.
+The `:text`, `:above`, `:below`, `:in`, and `:not-in` modifiers are not supported in this version of the IBM FHIR server and use of these modifiers will result in an HTTP 400 error with an OperationOutcome that describes the failure.
```
Comment (lmsurpre) [inline docs/src/pages/Conformance.md:177]: Totally unrelated to your change, but I just noticed that we say `:above` and `:below` aren't supported, but we do actually support those for URI types...we just don't support those on token types.  Maybe we can reword this to be more clear?
```suggestion
The `:text`, `:above`, `:below`, `:in`, and `:not-in` token modifiers are not yet supported by the IBM FHIR server and use of these modifiers will result in an HTTP 400 error with an OperationOutcome that describes the failure.
```

The `:text` modifier actually applies to String types too, and we don't support it there either, but I think thats just too much information to distinguish...they can infer it from the table above.

Comment (tbieste) [reply docs/src/pages/Conformance.md]: Reworded.

File: fhir-search/src/main/java/com/ibm/fhir/search/parameters/QueryParameterValue.java [hunk 2]
```diff
@@ -150,6 +151,22 @@ public void setHidden(boolean hidden) {
...
+        return ofTypeModifier;
+    }
+
+    /**
+     * Gets whether the value of an :of-type modifier.
+     * @param ofTypeModifier true if value of an :of-type modifier
+     */
+    public void setOfTypeModifier(boolean ofTypeModifier) {
+        this.ofTypeModifier = ofTypeModifier;
+    }
```
Comment (lmsurpre) [inline fhir-search/src/main/java/com/ibm/fhir/search/parameters/QueryParameterValue.java:168]: i think you mixed up the javadoc on these

File: fhir-search/src/main/java/com/ibm/fhir/search/parameters/QueryParameterValue.java [hunk 3]
```diff
@@ -190,7 +208,7 @@ public String toString() {
                     throw new IllegalStateException("Components of a composite search parameter may only have a single value");
                 }
                 returnString.append(componentDelim).append(componentValues.get(0));
-                componentDelim = "$";
+                componentDelim = ofTypeModifier ? SearchConstants.PARAMETER_DELIMITER : SearchConstants.COMPOSITE_DELIMITER;
```
Comment (lmsurpre) [inline fhir-search/src/main/java/com/ibm/fhir/search/parameters/QueryParameterValue.java:211]: where is QueryParameterValue.toString() used?
is it ok that this will produce `type.system|type.code|value` and not `type.system|type.code$value`?

Comment (tbieste) [reply fhir-search/src/main/java/com/ibm/fhir/search/parameters/QueryParameterValue.java:211]: It is used when generating the self/next/previous links for the search response Bundle.

Comment (lmsurpre) [reply fhir-search/src/main/java/com/ibm/fhir/search/parameters/QueryParameterValue.java:211]: awesome, then the use of `|` to match what came in makes perfect sense

Comment (tbieste) [reply fhir-search/src/main/java/com/ibm/fhir/search/parameters/QueryParameterValue.java:211]: I'll update the comment to be more explicit about that.

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java [hunk 4]
```diff
@@ -1281,13 +1285,41 @@ private static void checkSearchParameterCombinations(Class<?> resourceType, List
...
+                        QueryParameter typeParameter = new QueryParameter(Type.TOKEN, SearchUtil.makeCompositeSubCode(ofTypeParmName, SearchConstants.OF_TYPE_MODIFIER_COMPONENT_TYPE),
+                            null, null, Collections.singletonList(typeParameterValue));
+                        parameterValue.addComponent(typeParameter);
+
+                        QueryParameterValue valueParameterValue = new QueryParameterValue();
+                        valueParameterValue.setValueCode(unescapeSearchParm(parts[parts.length - 1]));
+                        QueryParameter valueParameter = new QueryParameter(Type.TOKEN, SearchUtil.makeCompositeSubCode(ofTypeParmName, SearchConstants.OF_TYPE_MODIFIER_COMPONENT_VALUE),
+                            null, null, Collections.singletonList(valueParameterValue));
+                        parameterValue.addComponent(valueParameter);
+                    } else {
```
Comment (michaelwschroeder) [inline fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:1315]: Might be ok if system was not specified, but if number of parts is not 2 or 3 I think we should definitely fail it.

File: fhir-search/src/main/java/com/ibm/fhir/search/parameters/QueryParameterValue.java (hunk 3 above)
Comment (tbieste) [reply fhir-search/src/main/java/com/ibm/fhir/search/parameters/QueryParameterValue.java:213]: Updated.

File: fhir-search/src/main/java/com/ibm/fhir/search/parameters/QueryParameterValue.java [hunk 5]
```diff
@@ -150,6 +151,22 @@ public void setHidden(boolean hidden) {
         this.hidden = hidden;
     }
 
+    /**
+     * Sets whether the value of an :of-type modifier.
```
[the remark this answers is not in the document]
Comment (tbieste) [reply fhir-search/src/main/java/com/ibm/fhir/search/parameters/QueryParameterValue.java]: Fixed.

File: fhir-search/src/main/java/com/ibm/fhir/search/parameters/QueryParameterValue.java (hunk 2 above)
Comment (tbieste) [reply fhir-search/src/main/java/com/ibm/fhir/search/parameters/QueryParameterValue.java]: Fixed.

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java (hunk 4 above)
Comment (tbieste) [reply fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:1315]: Yeah, this was the one part I wasn't sure of.  Do we fail if not all these parts are provided?  It feels like allowing 2 parts would be ok, and then fail or no results if just one.  Trying to be consistent with other token searches, too.
@lmsurpre Do you have a preference?

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java [hunk 6]
```diff
@@ -1281,13 +1285,41 @@ private static void checkSearchParameterCombinations(Class<?> resourceType, List
...
                  * "For token parameters on elements of type ContactPoint, uri, or boolean,
                  * the presence of the pipe symbol SHALL NOT be used - only the
                  * [parameter]=[code] form is allowed
                  */
                 String[] parts = v.split(SearchConstants.BACKSLASH_NEGATIVE_LOOKBEHIND + "\\|");
+                if (Modifier.OF_TYPE.equals(modifier)) {
+                    // Convert :of-type into a composite search parameter
+                    final String ofTypeParmName = searchParameter.getCode().getValue() + SearchConstants.OF_TYPE_MODIFIER_SUFFIX;
+                    parameterValue.setOfTypeModifier(true);
+                    if (parts.length > 1 && parts.length < 4) {
```
[the remark this answers is not in the document]
Comment (tbieste) [reply fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:1300]: Will resolve this based on resolution on comment below.

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java (hunk 4 above)
Comment (michaelwschroeder) [reply fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:1315]: I don't have a strong preference on 3 parts vs. 2 parts, but I think if only 1 part then we should fail or ignore based on 'lenient' setting.

Comment (lmsurpre) [reply fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:1315]: I don't have a strong preference on this one.  I'd be fine to require either 2 or 3 parts.
I would think it would have the following semantics:
* If they use 2 parts, it means the first part is a search on the Identifier.type.code (irregardless of system)
* And if they use 3 parts, I we allow them to leave the first part empty (which should be like searching on a system-less code)

Comment (michaelwschroeder) [reply fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:1315]: works for me

Comment (tbieste) [reply fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:1315]: I'll update it to require at least 2 parts.

File: fhir-search/src/main/java/com/ibm/fhir/search/parameters/QueryParameterValue.java [hunk 7]
```diff
@@ -208,6 +208,8 @@ public String toString() {
                     throw new IllegalStateException("Components of a composite search parameter may only have a single value");
                 }
                 returnString.append(componentDelim).append(componentValues.get(0));
+                // Since the self/next/previous links generated by UriBuilder for the search response Bundle use this output,
+                // this needs to use "|" instead of "$" as the component delimiter
```
Comment (lmsurpre) [inline fhir-search/src/main/java/com/ibm/fhir/search/parameters/QueryParameterValue.java:212]: good useful comment

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java (hunk 4 above)
Comment (lmsurpre) [reply fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:1315]: > I'll update it to require at least 2 parts.
I thought it already did :-)

Comment (tbieste) [reply fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:1315]: If there's only 1 part, then line 1316 still produces a valid search against just the value (without any system or code).

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/JDBCParameterBuildingVisitor.java [hunk 8]
```diff
@@ -587,12 +623,42 @@ public boolean visit(java.lang.String elementName, int elementIndex, Identifier
...
+                        // Create as composite parm since type/value are correlated for the :of-type modifier
+                        CompositeParmVal cp = new CompositeParmVal();
+                        cp.setResourceType(resourceType);
+                        cp.setName(searchParamCode + SearchConstants.OF_TYPE_MODIFIER_SUFFIX);
+
+                        // type
+                        p = new TokenParmVal();
+                        p.setResourceType(cp.getResourceType());
+                        p.setName(SearchUtil.makeCompositeSubCode(cp.getName(), SearchConstants.OF_TYPE_MODIFIER_COMPONENT_TYPE));
+                        if (typeCoding.getSystem() != null) {
```
[the remark this answers is not in the document]
Comment (tbieste) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/JDBCParameterBuildingVisitor.java:645]: Since we will require search query to only provide 2 parts (code|value), it will still be able to find codes without an explicit system defined.

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java (hunk 4 above)
Comment (tbieste) [reply fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:1318]: Updates made.  Should be ready to merge after successful build.

Comment (michaelwschroeder) [reply fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:1318]: I think if the :of-type modifier is specified, then anything other than 2 or 3 parts should produce an error. Think this case should be handled same as line 1300

Comment (tbieste) [reply fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:1318]: I can have it do that.  It will be strange that all other token search don't check for too many parts.  For example, you can search for "identifier=system|code|other1|other2|other3" and there's no error.

Comment (michaelwschroeder) [reply fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:1318]: I think all other token searches should generate an error too :). But for consistency's sake, I can live with it as is.

**Review by michaelwschroeder — APPROVED** [review]
LGTM

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java (hunk 4 above)
Comment (tbieste) [reply fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:1318]: Sounds good. I'll leave it for consistency's sake.  If we want to do stricter parsing of token searches and throw errors that can be done consistently at a later time.

**Review by lmsurpre — APPROVED** [review]
LGTM
